### PR TITLE
Changed terrain for central island, added weighted nets, rebalanced costs for elementalist spells.

### PR DESCRIPTION
**Pull request by HallerJohn**
Changed mid island
new item - hunting net
ward recipe changed
had to redo ele mana changes but it is still the same as previous pr

### CODE REVIEW

File: wurst/__legacy__/lib/ID.jurst [hunk 1]
```diff
@@ -408,6 +408,7 @@ library ID requires ObjectIdGenerator
 		public constant integer SPELL_QM_REFINED_POISON_SPEAR = 'A0JB'
 		public constant integer SPELL_QM_ULTRA_POISON_SPEAR   = 'A0JC'
 		public constant integer SPELL_QM_NETS                 = 'A0JD'
+		public constant integer SPELL_QM_HUNTING_NET		  = 'A0JM'
```
Comment (jonathanmclaus) [inline wurst/__legacy__/lib/ID.jurst]: Please align these definitions.

File: wurst/_items/HuntingNet.wurst [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    ..setHitPoints(75)
+    ..setIgnoreCooldown(false)
+    ..setNumberofCharges(1)
+    ..setPerishable(true)
+    ..setCooldownGroup("Aens")
+ 
+function moveTarget()
+    let caster = GetSpellAbilityUnit()
+    let target = GetSpellTargetUnit()
+    if target.isType(UNIT_TYPE_FLYING) or targets.has(target.getTypeId())
```
Comment (jonathanmclaus) [inline wurst/_items/HuntingNet.wurst]: Can you not check for acquatic units? Amphibious units shouldn't be moved, but all acquatic ones should be.

File: wurst/_items/HuntingNet.wurst [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        target.setPos(caster.getPos())
+    
+function preventNet()
```
Comment (jonathanmclaus) [inline wurst/_items/HuntingNet.wurst]: Can you not re-use this from the existing code?

File: wurst/_items/HuntingNet.wurst [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    let caster = GetSpellAbilityUnit()
+    if target.isTroll() and (target.getAbilityLevel(BUFF_ANTI_MAGIC_SHIELD) > 0 or target.getAbilityLevel(BUFF_ENSNARE_GROUND) > 0) and caster.getTypeId() != UNIT_DISCO_DUCK and caster.getTypeId() != UNIT_ONE
+        caster.issueImmediateOrder("stop")
+        simError(caster.getOwner(), "Cannot ensnare target with Anti-Magic shield")
+        
+function onHuntNet()
+    let target = GetSpellTargetUnit()
+    let pos = target.getPos()
+    if target.isTroll()
+        doPeriodicallyTimed(0.1, 2.5) netCounter ->
```
Comment (jonathanmclaus) [inline wurst/_items/HuntingNet.wurst]: Use the definition of `HERO_DURATION` from above. Do not use magic numbers that are not immediately obvious.

I'd also recommend not using `0.1` in favor of `ANIMATION_PERIOD`, but I don't care as much if this is temporary.

File: wurst/_items/HuntingNet.wurst [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            if not target.isAlive()
+                createItem(ITEM_HUNTING_NET, pos)
+                netCounter.stop()
+    else
+        doPeriodicallyTimed(0.1, 13.0) netCounter ->
```
Comment (jonathanmclaus) [inline wurst/_items/HuntingNet.wurst]: Same as above - use `NORMAL_DURATION`. And I think standard practice is not use the extra `0`, e.g. `13.` not `13.0`.

**Comment by jonathanmclaus** [comment]
In general, this could have been 3 PRs - the files you changed are entirely separate AFAIK. I will accept regardless though.

File: wurst/_items/HuntingNet.wurst (hunk 5 above)
Comment (HallerJohn) [reply wurst/_items/HuntingNet.wurst]: i had a problem where it was reading the variable as a string so i just added . to force it to read as a number

Comment (HallerJohn) [reply wurst/_items/HuntingNet.wurst]: also do periodically timed didnt work when i used hero-duration

File: wurst/_items/HuntingNet.wurst (hunk 3 above)
Comment (HallerJohn) [reply wurst/_items/HuntingNet.wurst]: i thought quazz said i couldnt base item abilities off custom abilities and to just add in code from other places to mimick that

File: wurst/_items/HuntingNet.wurst (hunk 2 above)
Comment (HallerJohn) [reply wurst/_items/HuntingNet.wurst]: im not sure, i used the code that checks for a ensnare abilities

File: wurst/_items/HuntingNet.wurst (hunk 3 above)
Comment (theQuazz) [reply wurst/_items/HuntingNet.wurst]: @HallerJohn you can reuse the code if possible

File: wurst/_items/HuntingNet.wurst (hunk 5 above)
Comment (jonathanmclaus) [reply wurst/_items/HuntingNet.wurst]: You need the `.` to force it from an integer to a float. You don't need the `0` though.

What is the error if you try using `HERO_DURATION`?

File: wurst/_items/HuntingNet.wurst [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+import ClosureEvents
+import ClosureTimers
+import ObjectIds
+import UnitExtensions
+import HashSet
+import SimError
+
+constant ABILITY_DUMMY_ID = compiletime(ABIL_ID_GEN.next())
+let BUFF_HUNTED = 'Hntd'
+let netDuration = 0.0
```
Comment (jonathanmclaus) [inline wurst/_items/HuntingNet.wurst]: This is unused. Please remove.

File: wurst/_items/HuntingNet.wurst (hunk 5 above)
Comment (HallerJohn) [reply wurst/_items/HuntingNet.wurst]: well its working now so it must have been some other thing i messed up before

File: wurst/_items/HuntingNet.wurst (hunk 2 above)
Comment (jonathanmclaus) [reply wurst/_items/HuntingNet.wurst]: Use the existing code for this too.

File: wurst/_items/HuntingNet.wurst [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        doPeriodicallyTimed(0.1, 13.0) netCounter ->
+            if not target.isAlive()
+                createItem(ITEM_HUNTING_NET, pos)
+                netCounter.stop()
+    
+init
+    registerSpellEffectEvent(ABILITY_DUMMY_ID) ->
+        preventNet()
+        moveTarget()
+        onHuntNet()
```
Comment (jonathanmclaus) [inline wurst/_items/HuntingNet.wurst]: Add line break at end.

File: wurst/__legacy__/lib/ID.jurst [hunk 8]
```diff
@@ -994,6 +995,7 @@ library ID requires ObjectIdGenerator
 		public constant integer ITEM_EMP             = 'I03M'
 		public constant integer ITEM_FIRE_BOMB       = 'I04J'
 		public constant integer ITEM_NETS            = 'I024'
+		public constant integer ITEM_HUNTING_NET	 = 'I060'
```
Comment (jonathanmclaus) [inline wurst/__legacy__/lib/ID.jurst]: Get rid of tab. You can probably configure VS Code to replace tabs with 2 or 4 spaces.

File: wurst/_items/HuntingNet.wurst [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    ..setNumberofCharges(1)
+    ..setPerishable(true)
+    ..setCooldownGroup("Aens")
+ 
+        
+function onHuntNet()
+    let target = GetSpellTargetUnit()
+    let pos = target.getPos()
+    real NET_DURATION = 0.
+    if target.isTroll()
```
Comment (jonathanmclaus) [inline wurst/_items/HuntingNet.wurst]: Use ternary operator.

File: wurst/_items/HuntingNet.wurst [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        NET_DURATION = NORMAL_DURATION
+    doPeriodicallyTimed(ANIMATION_PERIOD, NET_DURATION) netCounter ->
+        if not target.isAlive()
+            createItem(ITEM_HUNTING_NET, pos)
+            netCounter.stop()
+    
+    
+init
+    registerSpellEffectEvent(HUNTING_NET_ENSNARE) ->
+        preventNet()
```
Comment (jonathanmclaus) [inline wurst/_items/HuntingNet.wurst]: Add a `TODO` comment stating that `moveTarget` needs to be conditional based on `preventNet`. Same for `onHuntNet` - attempting to net an AMS unit and then killing will probably get you a free net currently.

File: wurst/_items/HuntingNet.wurst [hunk 11]
```diff
@@ -76,18 +76,16 @@ function onHuntNet()
     let target = GetSpellTargetUnit()
     let pos = target.getPos()
     real NET_DURATION = 0.
-    if target.isTroll()
-        NET_DURATION = HERO_DURATION
-    else 
-        NET_DURATION = NORMAL_DURATION
+    NET_DURATION = (target.isTroll())? HERO_DURATION : NORMAL_DURATION
```
Comment (jonathanmclaus) [inline wurst/_items/HuntingNet.wurst]: You don't need extra parenthesis.

`NET_DURATION = target.isTroll() ? HERO_DURATION : NORMAL_DURATION `.

**Comment by jonathanmclaus** [comment]
Did you start this branch prior to your last change for the map file?

**Comment by HallerJohn** [comment]
I started this branch after i originally made those 2 fire spots near mammy that didnt work

**Comment by HallerJohn** [comment]
also dont worry about being a stickler, if i want my code in the game it needs to be to the standard of the rest of it. also its good practice for me to write clean efficient code

**Comment by HallerJohn** [comment]
did i screw something up with how i made the branch?

**Comment by jonathanmclaus** [comment]
No. You just need to manage the merge manually.

**Comment by HallerJohn** [comment]
how do i do that?

**Pull request merged**